### PR TITLE
Use loadbalancer in addsvc example

### PR DESCRIPTION
**Pull request by jupp0r**
This change makes changes to the addsvc example in order to demonstrate
usage of package loadbalancer. As prerequisite, endpoint handling is
abstracted from transport implementations. Transport specific factories
now generate functions for setup and teardown of connections.

Resolves issue #89

### CODE REVIEW

File: examples/addsvc/client/client.go [hunk 1]
```diff
@@ -1,23 +1,20 @@
-package grpc
+package main
```
Comment (peterbourgon) [inline examples/addsvc/client/client.go]: Whoops! Typo? :)

Comment (jupp0r) [inline examples/addsvc/client/client.go]: oops

File: examples/addsvc/client/grpc/factory.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	grpctransport "github.com/go-kit/kit/transport/grpc"
+)
+
+// SumEndpointFactory transforms GRPC host:port strings into Endpoints that call the Sum method on a GRPC server
+// at that address.
+func SumEndpointFactory(instance string) (endpoint.Endpoint, io.Closer, error) {
+	cc, err := grpc.Dial(instance, grpc.WithInsecure())
+	return grpctransport.NewClient(cc, "Add", "Sum", encodeSumRequest, decodeSumResponse, pb.SumReply{}).Endpoint(),
+		cc,
+		err
```
Comment (peterbourgon) [inline examples/addsvc/client/grpc/factory.go]: Would you mind reformatting like this?

```go
return grpctransport.NewClient(
	cc, 
	"Add", 
	"Sum", 
	encodeSumRequest, 
	decodeSumResponse, 
	pb.SumReply{},
).Endpoint(), cc, err
```

File: examples/addsvc/client/grpc/factory.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+// ConcatEndpointFactory transforms GRPC host:port strings into Endpoints that call the Concat method on a GRPC server
+// at that address.
+func ConcatEndpointFactory(instance string) (endpoint.Endpoint, io.Closer, error) {
+	cc, err := grpc.Dial(instance, grpc.WithInsecure())
+	return grpctransport.NewClient(cc, "Add", "Concat", encodeConcatRequest, decodeConcatResponse, pb.ConcatReply{}).Endpoint(),
+		cc,
+		err
```
Comment (peterbourgon) [inline examples/addsvc/client/grpc/factory.go]: Likewise above.

File: examples/addsvc/client/httpjson/factory.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		return nil, nil, err
+	}
+	sumURL.Path = "/sum"
+
+	client := httptransport.NewClient(
+		"GET",
+		sumURL,
+		server.EncodeSumRequest,
+		server.DecodeSumResponse,
+		httptransport.SetClient(nil))
```
Comment (peterbourgon) [inline examples/addsvc/client/httpjson/factory.go]: Can you put the final ) on its own line please?

File: examples/addsvc/client/httpjson/factory.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+		return nil, nil, err
+	}
+	concatURL.Path = "/concat"
+
+	client := httptransport.NewClient(
+		"GET",
+		concatURL,
+		server.EncodeConcatRequest,
+		server.DecodeConcatResponse,
+		httptransport.SetClient(nil))
```
Comment (peterbourgon) [inline examples/addsvc/client/httpjson/factory.go]: Likewise above.

File: examples/addsvc/client/main.go [hunk 6]
```diff
@@ -143,3 +111,22 @@ func main() {
...
 	}
 }
+
+func seed() int64 {
+	b := make([]byte, 8)
+	_, err := rand.Read(b)
+	if err != nil {
+		fmt.Println(err)
+	}
+	return int64(binary.BigEndian.Uint64(b))
```
Comment (peterbourgon) [inline examples/addsvc/client/main.go]: Why not rand.NewInt63n?

File: examples/addsvc/client/main.go [hunk 7]
```diff
@@ -143,3 +111,22 @@ func main() {
...
+}
+
+func buildEndpoint(instances []string, factory loadbalancer.Factory, randomSeed int64, logger log.Logger) endpoint.Endpoint {
+	return loadbalancer.Retry(10, 10*time.Second,
+		loadbalancer.NewRandom(
+			static.NewPublisher(
+				instances,
+				factory,
+				logger),
+			randomSeed))
```
Comment (peterbourgon) [inline examples/addsvc/client/main.go]: Ah, you don't need to pass a random number as the seed.

Comment (peterbourgon) [inline examples/addsvc/client/main.go]: Suggest reformulating this as

```go
publisher := static.NewPublisher(instances, factory, logger)
random := loadbalancer.NewRandom(publisher, seed)
return loadbalancer.Retry(10, 10*time.Second, random)
```

File: examples/addsvc/client/main.go [hunk 8]
```diff
@@ -50,81 +51,48 @@ func main() {
...
+		sumFactory = thriftClient.SumEndpoint
+		concatFactory = thriftClient.ConcatEndpoint
 
 	default:
 		logger.Log("err", "invalid transport")
 		os.Exit(1)
 	}
 
+	sum := buildEndpoint(instances, sumFactory, randomSeed, logger)
+	concat := buildEndpoint(instances, concatFactory, randomSeed, logger)
```
Comment (peterbourgon) [inline examples/addsvc/client/main.go]: randomSeed can just be like `123` or `rand.Int63n`

File: examples/addsvc/server/encode_decode.go [hunk 9]
```diff
@@ -13,7 +13,7 @@ import (
 func DecodeSumRequest(r *http.Request) (interface{}, error) {
 	var request SumRequest
 	err := json.NewDecoder(r.Body).Decode(&request)
-	return request, err
+	return &request, err
```
Comment (peterbourgon) [inline examples/addsvc/server/encode_decode.go]: Can you quickly summarize why this change was necessary?

File: examples/addsvc/client/main.go (hunk 6 above)
Comment (jupp0r) [inline examples/addsvc/client/main.go]: There is only one request made by the client before it shuts down. When using a constant seed, the same endpoint is picked every time, so you cannot have two servers running, launch the client repeatedly and see the requests arriving randomly. Now that I think about it, seeding with a timestamp would be easier, though.

Comment (peterbourgon) [inline examples/addsvc/client/main.go]: I see. Yeah, time.Now().UnixNano() works.

File: examples/addsvc/server/encode_decode.go (hunk 9 above)
Comment (jupp0r) [inline examples/addsvc/server/encode_decode.go]: I discovered this during manual testing. Apparently it's broken on master.

    ts=2016-04-01T13:19:13Z caller=stdlib.go:89 file="http: panic serving [::1" msg="]:60046: interface conversion: interface {} is server.ConcatRequest, not *server.ConcatRequest"

Comment (peterbourgon) [inline examples/addsvc/server/encode_decode.go]: Damn. Thanks for the catch.

**Comment by jupp0r** [comment]
I addressed all comments, quashed commits and rebased to master.

**Comment by peterbourgon** [comment]
Nice. Thanks so much for the contribution!
